proj-android/PowerAuthLibrary/src/main/java/io/getlime/security/powerauth/sdk/PowerAuthTokenStore.java

```diff
@@ -0,0 +1,421 @@
+    /**
+     * Reference to {@link PA2Keychain} for persistent storage purposes
+     */
+    private final PA2Keychain keychain;
+    /**
+     * Reference to @{link PA2Client} for networking purposes
+     */
+    private final PA2Client httpClient;
+    /**
+     * A dictionary mapping token's name to private token's data. This is the in-memory cache
+     * which speeds up querying for tokens.
+     */
+    private final HashMap<String, PowerAuthPrivateTokenData> localTokens;
+    /**
+     * A encryptor created from server's master public key.
+     */
+    private final ECIESEncryptor encryptor;
+
+
+    /**
+     * Constructs a new token store with references to parent {@link PowerAuthSDK}, {@link PA2Keychain}
+     * as storage and {@link PA2Client} for networking.
+     *
+     * @param sdk a parent object which created this instance
+     * @param keychain a keychain as persistent storage
+     * @param httpClient a HTTP client for networking
+     */
+    public PowerAuthTokenStore(
+            @NonNull PowerAuthSDK sdk,
+            @NonNull PA2Keychain keychain,
+            @NonNull PA2Client httpClient) {
+        this.sdk = sdk;
+        this.keychain = keychain;
+        this.httpClient = httpClient;
+        this.localTokens = new HashMap<>();
+        this.encryptor = new ECIESEncryptor(httpClient.getConfiguration().getMasterServerPublicKey(), null);
+    }
+
+
+    /**
+     * @return true if this instance can provide {@link PowerAuthToken} objects.
+     */
+    public synchronized boolean canRequestForAccessToken() {
+        return sdk != null && sdk.hasValidActivation();
+    }
+
+    /**
+     * Create a new access token with given name for requested signature factors.
+     * <p>
+     * Note that the method is thread safe, but it's not recommended to request for the same token
+     * name in parallel when the token is not created yet. If the method returns an asynchronous task,
+     * then the pending HTTP request to the server has been issued, so you should not ask for
+     * the same token while the task is in processing. You can use {@code PowerAuthTokenStore.hasLocalToken()} method
+     * to check, whether the token is already in the local database.
+     *
+     * @param context Context
+     * @param tokenName Name of requested token.
+     * @param authentication An authentication instance specifying what factors should be used for token creation.
+     * @param listener Listener with callbacks to receive a token.
+     * @return {@code AsyncTask} associated with the running server request or null if request has been processed synchronously.
+     */
+    public @Nullable AsyncTask requestAccessToken(@NonNull final Context context, @NonNull final String tokenName, @NonNull PowerAuthAuthentication authentication, @NonNull final IGetTokenListener listener) {
+
+        Throwable error = null;
+        PowerAuthPrivateTokenData tokenData = null;
+
+        synchronized (this) {
```

Comment:
Note: I admit it is probably my lack of knowledge here, but why do we need to synchronize this code here? Result of call to "canRequestForAccessToken" may be invalidated from outside this method, call to "getTokenData" may fail despite this synchronization. Or are we trying to synchronize only call to "getTokenData" here?

Reply:
You're correct, but if you invalidate the session during the getting token from the server, then the created token still cannot be used afterwards. That's OK and predictable behaviour. So, I need that lock only during accessing the private data structures (e.g. hash map). Despite of that, your hint had a good point and as a result, I've improved the code a little bit. Now the token is not stored to the keychain when is received from the server and the SDK has no longer a valid activation.